File: pnpm11/config/reader/src/osProxy.ts
import { execFileSync } from 'node:child_process'

/**
 * Proxy URLs and bypass list read from the operating system.
 *
 * Strings are `host:port` when the system settings omit the URL scheme.
 * Callers prefix `http://` when they build a request agent.
 */
export interface OsProxy {
  httpsProxy?: string
  httpProxy?: string
  noProxy?: string
}

export interface ProxySourceInput {
  httpsProxy: unknown
  httpProxy: unknown
  proxy: unknown
  noProxy: unknown
  noproxy: unknown
  envHttpsProxy: string | undefined
  envHttpProxy: string | undefined
  envProxy: string | undefined
  envNoProxy: string | undefined
  os: OsProxy
}

export interface ResolvedProxySettings {
  httpsProxy: unknown
  httpProxy: unknown
  noProxy: unknown
}

/**
 * Config, then the environment, then the operating system.
 *
 * An empty environment variable is a value: it shadows the operating
 * system and resolves to no proxy. `proxy: false` does not fall through.
 * HTTPS filled only from the operating system does not hide a distinct
 * OS HTTP proxy.
 */
export function resolveProxyFromSources (input: ProxySourceInput): ResolvedProxySettings {
  const httpsWasUnset = !input.httpsProxy
  const legacyProxy = input.proxy === '' ? undefined : input.proxy
  const httpsFromConfigOrEnv = httpsWasUnset
    ? legacyProxy ?? input.envHttpsProxy
    : input.httpsProxy
  const httpsProxy = httpsWasUnset
    ? httpsFromConfigOrEnv ?? input.os.httpsProxy
    : input.httpsProxy

  let httpProxy = input.httpProxy
  if (!httpProxy) {
    const envHttpProxy = input.envHttpProxy ?? input.envProxy
    const httpFromUpper = httpsFromConfigOrEnv ?? envHttpProxy
    if (httpFromUpper != null && httpFromUpper !== '') {
      httpProxy = httpFromUpper
    } else if (httpFromUpper === '') {
      httpProxy = ''
    } else {
      const httpsFromOs = httpsProxy != null && httpsProxy !== '' && httpsProxy !== httpsFromConfigOrEnv
        ? httpsProxy
        : undefined
      httpProxy = input.os.httpProxy ?? httpsFromOs
    }
  }

  let noProxy = input.noProxy
  if (!noProxy) {
    const fromUpper = input.noproxy ?? input.envNoProxy
    noProxy = fromUpper ?? input.os.noProxy
  }

  return { httpsProxy, httpProxy, noProxy }
}

/**
 * The operating-system proxy for this process.
 *
 * Jest sets `JEST_WORKER_ID`. Tests that need a specific OS proxy pass it
 * in, so discovery here stays empty and does not read the machine.
 */
export function currentOsProxy (): OsProxy {
  if (process.env.JEST_WORKER_ID != null) return {}
  return readOsProxy()
}

function readOsProxy (): OsProxy {
  if (process.platform === 'win32') return readWindowsProxy()
  if (process.platform === 'darwin') return readMacosProxy()
  return {}
}

function readWindowsProxy (): OsProxy {
  const stdout = commandStdout('reg', [
    'query',
    'HKEY_CURRENT_USER\\Software\\Microsoft\\Windows\\CurrentVersion\\Internet Settings',
  ])
  return stdout == null ? {} : parseWindowsInternetSettings(stdout)
}

function readMacosProxy (): OsProxy {
  const stdout = commandStdout('scutil', ['--proxy'])
  return stdout == null ? {} : parseMacosScutilProxy(stdout)
}

function commandStdout (program: string, args: string[]): string | undefined {
  try {
    return execFileSync(program, args, {
      encoding: 'utf8',
      timeout: 5000,
      windowsHide: true,
    })
  } catch {
    return undefined
  }
}

/**
 * Parse `reg query` output for the Internet Settings key.
 *
 * `ProxyEnable` of `0` or a missing value means no proxy, even when
 * `ProxyServer` is present. A server value without `=` applies to both
 * schemes. `http=` and `https=` entries are read on their own.
 */
export function parseWindowsInternetSettings (output: string): OsProxy {
  if (registryDword(output, 'ProxyEnable') === 0) return {}
  const server = registryValue(output, 'ProxyServer')?.trim() ?? ''
  const { httpProxy, httpsProxy } = splitProxyServer(server)
  const noProxy = joinBypassEntries((registryValue(output, 'ProxyOverride') ?? '').split(';'))
  return {
    httpProxy,
    httpsProxy,
    noProxy: noProxy === '' ? undefined : noProxy,
  }
}

/**
 * Parse `scutil --proxy` output.
 *
 * A scheme is used only when its enable flag is non-zero and a host is
 * present. `ExceptionsList` becomes the bypass list.
 */
export function parseMacosScutilProxy (output: string): OsProxy {
  let httpEnable = false
  let httpsEnable = false
  let httpHost: string | undefined
  let httpsHost: string | undefined
  let httpPort: number | undefined
  let httpsPort: number | undefined
  const exceptions: string[] = []
  let inExceptions = false

  for (const line of output.split('\n')) {
    const trimmed = line.trim()
    if (inExceptions) {
      if (trimmed === '}') {
        inExceptions = false
        continue
      }
      const separator = trimmed.indexOf(' : ')
      if (separator !== -1) {
        const entry = normalizeBypassEntry(trimmed.slice(separator + 3).trim())
        if (entry != null) exceptions.push(entry)
      }
      continue
    }
    if (trimmed.startsWith('ExceptionsList') && trimmed.includes('<array>')) {
      inExceptions = true
      continue
    }
    const separator = trimmed.indexOf(' : ')
    if (separator === -1) continue
    const key = trimmed.slice(0, separator).trim()
    const value = trimmed.slice(separator + 3).trim()
    switch (key) {
      case 'HTTPEnable':
        httpEnable = flagEnabled(value)
        break
      case 'HTTPSEnable':
        httpsEnable = flagEnabled(value)
        break
      case 'HTTPProxy':
        httpHost = nonEmpty(value)
        break
      case 'HTTPSProxy':
        httpsHost = nonEmpty(value)
        break
      case 'HTTPPort':
        httpPort = positivePort(value)
        break
      case 'HTTPSPort':
        httpsPort = positivePort(value)
        break
    }
  }

  return {
    httpProxy: enabledEndpoint(httpEnable, httpHost, httpPort),
    httpsProxy: enabledEndpoint(httpsEnable, httpsHost, httpsPort),
    noProxy: exceptions.length === 0 ? undefined : exceptions.join(','),
  }
}

function registryValue (output: string, name: string): string | undefined {
  for (const line of output.split('\n')) {
    const value = registryLineValue(line, name)
    if (value != null) return value
  }
  return undefined
}

function registryLineValue (line: string, name: string): string | undefined {
  if (!line.startsWith('    ')) return undefined
  const rest = line.slice(4)
  if (rest.length < name.length || rest.slice(0, name.length).toLowerCase() !== name.toLowerCase()) {
    return undefined
  }
  if (!rest.slice(name.length).startsWith('    ')) return undefined
  const afterName = rest.slice(name.length + 4)
  const typeEnd = afterName.indexOf('    ')
  if (typeEnd === -1) return undefined
  return afterName.slice(typeEnd + 4).trim()
}

function registryDword (output: string, name: string): number {
  const raw = registryValue(output, name)
  if (raw == null) return 0
  const hex = raw.startsWith('0x') ? raw.slice(2) : raw
  const parsed = Number.parseInt(hex, 16)
  return Number.isNaN(parsed) ? 0 : parsed
}

function splitProxyServer (server: string): { httpProxy?: string, httpsProxy?: string } {
  if (server === '') return {}
  if (!server.includes('=')) {
    return { httpProxy: server, httpsProxy: server }
  }
  let httpProxy: string | undefined
  let httpsProxy: string | undefined
  for (const entry of server.split(';')) {
    const separator = entry.indexOf('=')
    if (separator === -1) continue
    const scheme = entry.slice(0, separator).trim().toLowerCase()
    const value = entry.slice(separator + 1).trim()
    if (value === '') continue
    if (scheme === 'http') httpProxy = value
    if (scheme === 'https') httpsProxy = value
  }
  return { httpProxy, httpsProxy }
}

function joinBypassEntries (entries: string[]): string {
  return entries.map(normalizeBypassEntry).filter((entry): entry is string => entry != null).join(',')
}

function normalizeBypassEntry (entry: string): string | undefined {
  const trimmed = entry.trim()
  if (trimmed === '' || trimmed.toLowerCase() === '<local>') return undefined
  const stripped = trimmed.startsWith('*.') ? trimmed.slice(2) : trimmed
  return stripped === '' ? undefined : stripped
}

function flagEnabled (value: string): boolean {
  return value !== '' && value !== '0'
}

function nonEmpty (value: string): string | undefined {
  return value === '' ? undefined : value
}

function positivePort (value: string): number | undefined {
  const port = Number(value)
  if (!Number.isInteger(port) || port <= 0 || port > 65535) return undefined
  return port
}

function enabledEndpoint (enabled: boolean, host: string | undefined, port: number | undefined): string | undefined {
  if (!enabled || host == null) return undefined
  return port == null ? host : `${host}:${port}`
}

File: pnpm/crates/config/src/os_proxy.rs
//! Operating-system proxy settings, the last source in the proxy cascade.
//!
//! Windows Internet Settings and the macOS system proxy fill a slot only
//! when no config layer and no environment variable named it. Linux has no
//! standard system proxy, so discovery there is empty.

#[cfg(all(not(test), any(windows, target_os = "macos")))]
use std::process::Command;

/// Proxy URLs and bypass list read from the operating system.
///
/// Strings are schemeless `host:port` when the system settings store them
/// that way. The network layer prefixes `http://` at parse time.
#[derive(Debug, Default, Clone, PartialEq, Eq)]
pub(crate) struct OsProxy {
    pub https: Option<String>,
    pub http: Option<String>,
    pub bypass: Option<String>,
}

/// Read the operating-system proxy, or an empty value when none is set.
///
/// Unit tests always get an empty value so cascade tests do not depend on
/// the machine running them. A missing `reg` or `scutil`, or a command
/// that fails, is also an empty value: proxy discovery must not fail an
/// install.
#[must_use]
pub(crate) fn discover_os_proxy() -> OsProxy {
    #[cfg(test)]
    {
        OsProxy::default()
    }
    #[cfg(not(test))]
    {
        read_os_proxy()
    }
}

#[cfg(not(test))]
fn read_os_proxy() -> OsProxy {
    #[cfg(windows)]
    {
        command_stdout(
            "reg",
            &[
                "query",
                r"HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Internet Settings",
            ],
        )
        .map(|output| parse_windows_internet_settings(&output))
        .unwrap_or_default()
    }
    #[cfg(target_os = "macos")]
    {
        command_stdout("scutil", &["--proxy"])
            .map(|output| parse_macos_scutil_proxy(&output))
            .unwrap_or_default()
    }
    #[cfg(not(any(windows, target_os = "macos")))]
    {
        OsProxy::default()
    }
}

#[cfg(all(not(test), any(windows, target_os = "macos")))]
fn command_stdout(program: &str, args: &[&str]) -> Option<String> {
    let output = Command::new(program)
        .args(args)
        .output()
        .ok()?;
    if !output.status.success() {
        return None;
    }
    Some(String::from_utf8_lossy(&output.stdout).into_owned())
}

/// Parse `reg query` output for the Internet Settings key.
///
/// `ProxyEnable` of `0` or a missing value means no proxy, even when
/// `ProxyServer` is present. A server value without `=` applies to both
/// schemes. `http=` and `https=` entries are read on their own; other
/// schemes are ignored.
#[cfg(any(test, windows))]
#[must_use]
pub(crate) fn parse_windows_internet_settings(output: &str) -> OsProxy {
    if registry_dword(output, "ProxyEnable").unwrap_or(0) == 0 {
        return OsProxy::default();
    }
    let (http, https) = split_proxy_server(
        registry_value(output, "ProxyServer")
            .as_deref()
            .unwrap_or("")
            .trim(),
    );
    let bypass = registry_value(output, "ProxyOverride")
        .map(|raw| join_bypass_entries(raw.split(';')))
        .filter(|value| !value.is_empty());
    OsProxy { https, http, bypass }
}

/// Parse `scutil --proxy` output.
///
/// A scheme is used only when its enable flag is non-zero and a host is
/// present. `ExceptionsList` becomes the bypass list.
#[cfg(any(test, target_os = "macos"))]
#[must_use]
pub(crate) fn parse_macos_scutil_proxy(output: &str) -> OsProxy {
    let mut http_enable = false;
    let mut https_enable = false;
    let mut http_host = None;
    let mut https_host = None;
    let mut http_port = None;
    let mut https_port = None;
    let mut exceptions = Vec::new();
    let mut in_exceptions = false;

    for line in output.lines() {
        let trimmed = line.trim();
        if in_exceptions {
            if trimmed == "}" {
                in_exceptions = false;
                continue;
            }
            if let Some((_, value)) = trimmed.split_once(" : ")
                && let Some(entry) = normalize_bypass_entry(value.trim())
            {
                exceptions.push(entry);
            }
            continue;
        }
        if trimmed.starts_with("ExceptionsList") && trimmed.contains("<array>") {
            in_exceptions = true;
            continue;
        }
        let Some((key, value)) = trimmed.split_once(" : ") else {
            continue;
        };
        let value = value.trim();
        match key.trim() {
            "HTTPEnable" => http_enable = flag_enabled(value),
            "HTTPSEnable" => https_enable = flag_enabled(value),
            "HTTPProxy" => http_host = non_empty(value),
            "HTTPSProxy" => https_host = non_empty(value),
            "HTTPPort" => http_port = positive_port(value),
            "HTTPSPort" => https_port = positive_port(value),
            _ => {}
        }
    }

    let bypass = (!exceptions.is_empty()).then(|| exceptions.join(","));
    OsProxy {
        http: enabled_endpoint(http_enable, http_host, http_port),
        https: enabled_endpoint(https_enable, https_host, https_port),
        bypass,
    }
}

#[cfg(any(test, windows))]
fn registry_value(output: &str, name: &str) -> Option<String> {
    output.lines().find_map(|line| registry_line_value(line, name))
}

/// A `reg query` line is `    <name>    <type>    <data>` with four-space separators.
#[cfg(any(test, windows))]
fn registry_line_value(line: &str, name: &str) -> Option<String> {
    let rest = line.strip_prefix("    ")?;
    if rest.len() < name.len() || !rest[..name.len()].eq_ignore_ascii_case(name) {
        return None;
    }
    let after_name = rest[name.len()..].strip_prefix("    ")?;
    let type_end = after_name.find("    ")?;
    Some(after_name[type_end + 4..].trim().to_string())
}

#[cfg(any(test, windows))]
fn registry_dword(output: &str, name: &str) -> Option<u32> {
    let raw = registry_value(output, name)?;
    let hex = raw.strip_prefix("0x").unwrap_or(&raw);
    u32::from_str_radix(hex, 16).ok()
}

#[cfg(any(test, windows))]
fn split_proxy_server(server: &str) -> (Option<String>, Option<String>) {
    if server.is_empty() {
        return (None, None);
    }
    if !server.contains('=') {
        let url = Some(server.to_string());
        return (url.clone(), url);
    }
    let mut http_proxy = None;
    let mut https_proxy = None;
    for entry in server.split(';') {
        let Some((scheme, value)) = entry.split_once('=') else {
            continue;
        };
        let value = value.trim();
        if value.is_empty() {
            continue;
        }
        match scheme
            .trim()
            .to_ascii_lowercase()
            .as_str()
        {
            "http" => http_proxy = Some(value.to_string()),
            "https" => https_proxy = Some(value.to_string()),
            _ => {}
        }
    }
    (http_proxy, https_proxy)
}

#[cfg(any(test, windows))]
fn join_bypass_entries<'a>(entries: impl Iterator<Item = &'a str>) -> String {
    entries
        .filter_map(normalize_bypass_entry)
        .collect::<Vec<_>>()
        .join(",")
}

#[cfg(any(test, windows, target_os = "macos"))]
fn normalize_bypass_entry(entry: &str) -> Option<String> {
    let entry = entry.trim();
    if entry.is_empty() || entry.eq_ignore_ascii_case("<local>") {
        return None;
    }
    let stripped = entry.strip_prefix("*.").unwrap_or(entry);
    (!stripped.is_empty()).then(|| stripped.to_string())
}

#[cfg(any(test, target_os = "macos"))]
fn flag_enabled(value: &str) -> bool {
    !value.is_empty() && value != "0"
}

#[cfg(any(test, target_os = "macos"))]
fn non_empty(value: &str) -> Option<String> {
    (!value.is_empty()).then(|| value.to_string())
}

#[cfg(any(test, target_os = "macos"))]
fn positive_port(value: &str) -> Option<u16> {
    let port = value.parse::<u16>().ok()?;
    (port > 0).then_some(port)
}

#[cfg(any(test, target_os = "macos"))]
fn enabled_endpoint(enabled: bool, host: Option<String>, port: Option<u16>) -> Option<String> {
    if !enabled {
        return None;
    }
    let host = host?;
    Some(match port {
        Some(port) => format!("{host}:{port}"),
        None => host,
    })
}

#[cfg(test)]
mod tests;
